### PR TITLE
Enriching Hub Status to include Node info

### PR DESCRIPTION
**Pull request by krmahadevan**
Currently the end-point /grid/api/hub/status 
does not provide information about the nodes 
attached to it and also the busy/free status 
per browser flavor on each of the nodes.

Enriching the end-point to provide node info 
when invoked with the option 
/grid/api/hub/status?configuration=nodes

Sample payload as below:

```json
{
  "nodes": [
    {
      "Id": "http://192.168.1.6:5555",
      "browsers": [
        {
          "browser": "safari",
          "slots": {
            "busy": 0,
            "total": 1
          }
        },
        {
          "browser": "chrome",
          "slots": {
            "busy": 0,
            "total": 5
          }
        },
        {
          "browser": "firefox",
          "slots": {
            "busy": 0,
            "total": 5
          }
        }
      ]
    }
  ],
  "success": true
}
```

- [X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#step-6-sign-the-cla)

### CODE REVIEW

**Comment by krmahadevan** [comment]
ping @diemol @shs96c @barancev 

@shs96c - As discussed in the SeConf in BLR, I have enriched the `/grid/api/hub/status` end-point to optionally provide node details when queried. I will send a separate PR for the node side.

**Comment by luke-hill** [comment]
For my own clarity, would this provide essentially what `http://url-to-grid/get_node_config?node=nameofnode` does, but instead of being for 1 node it is for all nodes?

**Comment by krmahadevan** [comment]
@luke-hill,

* I am not aware of an end-point "/get_node_config" available in the Hub
that is currently doing this. Can you please help point me to it. I am
currently just enriching this
<https://github.com/SeleniumHQ/selenium/blob/master/java/server/src/org/openqa/grid/web/Hub.java#L134>
endpoint
* To answer to your question, no it does not because we aren't getting hold
of the configuration of the node (which is static), but this is dynamic
information and will reflect exactly how many sessions are in use vs free
per browser flavor on each of the nodes at any given point in time.

**Comment by luke-hill** [comment]
The top url I mentioned might be an SGE addon (Not sure).

Rightio. But sorry to be a pain. Isn't that provided (I hope this isn't SGE as well!), by `url-to-grid/grid/console`

**Comment by krmahadevan** [comment]
Yes the console ( url-to-grid/grid/console) provides you with this info.
But it's html and not JSON. So you would need to screen scrape to get this
data. This PR makes it easy to extract out that info.

> --

Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/

My Technical Scribbings @ http://rationaleemotions.wordpress.com/

**Comment by krmahadevan** [comment]
@mach6 - Thanks for the comments. I have extracted out all repeating strings in that class and converted them into private constants. Please take a look.

File: java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java [hunk 1]
```diff
@@ -167,4 +185,57 @@ protected void process(
       throw new IOException(e);
     }
   }
+
+  private static boolean IsKeyPresentIn(List<String> keys, String key) {
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java]: Nit: java method names begin with a lower-case letter. We can fix this once the PR is merged in.

File: java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java [hunk 2]
```diff
@@ -109,35 +124,38 @@ protected void process(
...
-    res.put("success", true);
+    res.put(SUCCESS, true);
 
     try {
         List<String> keysToReturn = null;
 
-        if (request.getParameter("configuration") != null && !"".equals(request.getParameter("configuration"))) {
-          keysToReturn = Arrays.asList(request.getParameter("configuration").split(","));
-        } else if (requestJSON != null && requestJSON.containsKey("configuration")) {
+        if (request.getParameter(CONFIGURATION) != null && !"".equals(request.getParameter(CONFIGURATION))) {
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java]: You can use `!Strings.isNullOrEmpty(request.getParameter(CONFIGURATION))` here for a slightly clearer check here (I realise that this wasn't done before)

File: java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java [hunk 3]
```diff
@@ -109,35 +124,38 @@ protected void process(
...
+          keysToReturn = Arrays.asList(request.getParameter(CONFIGURATION).split(","));
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java]: Are empty values allowed in the list? If not `Splitter.on(",").trimResults().omitEmptyStrings().splitToList()` may be a better fit.

File: java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java [hunk 4]
```diff
@@ -167,4 +185,57 @@ protected void process(
...
+  }
+
+  private List<Map<String, Object>> getNodesInfo() {
+    List<RemoteProxy> proxies = getRegistry().getAllProxies().getSorted();
+    return proxies.stream().map(this::getNodeInfo).collect(toList());
+  }
+
+  private Map<String, Object> getNodeInfo(RemoteProxy remoteProxy) {
+    return ImmutableSortedMap.of(
+        "Id", remoteProxy.getId(),
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java]: Why is `Id` capitalised, but `browser` not? Maybe leave both lower case?

File: java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java [hunk 5]
```diff
@@ -167,4 +185,57 @@ protected void process(
...
+    }
+    return BUSY;
+  }
+
+  private static String getBrowser(TestSlot slot) {
+    return slot.getCapabilities().get(CapabilityType.BROWSER_NAME).toString();
+  }
+
+  private static <T> Collector<T, ?, Integer> counting() {
+    return reducing(0, e -> 1, Integer::sum);
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/servlet/HubStatusServlet.java]: This looks like an IntStreams `sum` method, but presumably that isn't appropriate to use here

**Comment by krmahadevan** [comment]
@shs96c 

>  I'd ask for a test of some description to ensure that this continues working....

I have now added required tests for this changeset and also addressed all the other review feedback. Please help take a look.

**Comment by peterabbott** [comment]
This appears to be a breaking change when using an existing grid for < 3.14 as it is expected the enhanced metadata. A graceful fall back could have been nice as the update was getting pulled in transitively via other dependencies.

**Pull request merged**